In [2]:
from __future__ import print_function
import os
import sys
import numpy as np
import pylab as pl
import pandas as pd
import numpy as np

%pylab inline

from IPython.display import Image

Populating the interactive namespace from numpy and matplotlib


# Null hypothesis: The % of former prisoners suspectible to recidivism , i.e. getting convicted of a felony within 3 years of release is same or higher for candidates who participated in the program as for the control group, significance level p=0.05


##### 0 for control group, 1 for program group
 $H_0$ : $P_1$ - $P_0$ >= 0
 
 $H_a$ : $P_1$ - $P_0$ < 0

 $P_0 = 0.1, P_1= 0.117$
 
 
 $\alpha$=0.05

# Test 1: Z Test

In [2]:
alpha=0.05
# 0 for control group, 1 for program group
P_0=0.117
P_1=0.1

n_0= 409
n_1= 568


#calculate sample proportion

p  = ((P_0 * n_0) + (P_1 * n_1 )) / (n_0 + n_1)
print('Sample Proportion: ' + str(p))

#calculate the standard error

se = np.sqrt( p * (1-p) * (1.0/n_0 + 1.0/n_1))
print('Standard Error: '+ str(se))
z_3y = (P_0 - P_1)/se

z_3y

Sample Proportion: 0.107116683726
Standard Error: 0.0200556791612


0.84764020521993222

In [3]:
def report_result(p,a):
    print ('is the p value ' + 
           '{0:.2f} smaller than the critical value {1:.2f}?'.format(p,a))
    if p < a:
        print ("YES!")
    else: 
        print ("NO!")
    
    print ('the Null hypothesis is {}'.format(\
                            'rejected' if p < a  else 'not rejected') )

###### The z statistic for 0.85 is 0.8023

In [4]:
p_3y = 1 - 0.8023

report_result(p_3y, alpha)

is the p value 0.20 smaller than the critical value 0.05?
NO!
the Null hypothesis is not rejected


# Test 2: Chi Square Test

### Note: Function for chi square test was formulated using example here:

http://math.hws.edu/javamath/ryan/ChiSquare.html



In [5]:
def evalChisq(sample_values):
    total_sum = np.sum(sample_values)
    sum_between_rows = np.sum(sample_values, axis = 0)
    sum_within_rows = np.sum(sample_values, axis = 1)
    diagonal_product_1 = sample_values[0][0] * sample_values[1][1]
    diagonal_product_2 = sample_values[0][1] * sample_values[1][0]
    diagonal_difference_sq = (diagonal_product_1 - diagonal_product_2)**2
    
    return ((total_sum * diagonal_difference_sq )/ (sum_between_rows[0] * sum_between_rows[1] * sum_within_rows[0] * sum_within_rows[1]))

In [3]:
def evalChisq1(values):
    values = np.array(values)
    E = np.empty_like(values)
    #total number of categories
    for j in range(len(values[0])):
        # program and control , therefore 2
        for i in range(2):
            
            #sum along rows * sum along cols / total sum
            E[i][j] = ((values[i,:].sum() * values[:,j].sum()) / 
                        (values).sum())
    return ((values - E)**2 / E).sum()


In [8]:

Ntot = 977

sample_values = np.array([[0.1 * 568, 0.9 * 568],[0.117 * 409 , 0.883 * 409]])
#print (evalChisq(sample_values))


In [10]:
from scipy.stats import chi2_contingency
obs = np.array([[10, 10, 20], [20, 20, 20]])
chi2_contingency(sample_values, )
chi2_contingency(sample_values, correction=False)


(0.71849391750528868,
 0.39663837626861698,
 1,
 array([[  60.84227636,  507.15772364],
        [  43.81072364,  365.18927636]]))

In [7]:
# degree of freedom = (number of rows -1 )(number of columns -1 )
dof = (2-1) * (2-1)

# Observations and Conclusion:

p value at dof 1 and 0.05 significance level is 3.84

0.718493917505 < 3.84

which means p >= 0.05

Which means we fail to reject the null

# Contingency Table



|convicted of a felony      |     yes        | no          |                   |                   
|---------------------------|----------------|-------------|-------------------|
| test sample               | 0.1 * 568      | 0.9 * 568   |        568        |
| control sample            | 0.117 * 409    | 0.883 * 409 |        409        |
|                           |                |             |                   |
| total                     |   104.653      |   872.347   |        977        |